### DISCUSSION TITLE
SCIM filter and db implementation

### DISCUSSION THREAD

**Discussion by amir4895** [elimity-com/scim#158] [Package Questions]
Hi all ,
I'm  writing SCIM integration to OKTA ,
As far as i understand SCIM mechanism with OKTA:
https://developer.okta.com/docs/reference/scim/scim-20/#scim-user-operations

Once user is being provisioned (created or update or  delete) SCIM send "Get" & GetAll requests to check states, i'm using this repo and i'm seeing  in the logs few requests prior to the actual "create" / "replace "

getAll     /scim/Users?filter=userName%20eq%20%22Name%40%22&startIndex=1&count=100"
get       scim/Users/id"

as far as i understand the get all should find the specific user by unique key that is being configured in the IDP (okta here)
and if the user found the SCIM call for get to get the user detail if not calling create 

my question is: 
i need to query my db with the "filter " content from the getAll request in order to return the IDP the results 

I saw that in the ListRequestParams struct we set "filter" attribute with the content of the filter from the query params  but it's a complicated struct, and i didn't find easy way to use it ,  is there any recommended  package in Go that can "translate" the Expression into simple literals in String ?

for example "userName%20eq%20%22Name%40%22&"

left = username 
comparator = eq
right = name  

I searched over Github and didn't find something suitable 
tnx

**Comment by q-uint** [comment]
https://github.com/scim2/filter-parser provides parsing for filter expressions, `String` allows you to stringify expressions again.

**Reply by amir4895** [reply]
Not sure I follow, how I can use the String or the filter-parser
considering i'm having this function signature :
func (exp *scimUserHandler) GetAll(r *http.Request, params scim.ListRequestParams) (scim.Page, error) 
How can i use the parser? :

when i trying to use param.filter i don't see any String function. 

i was expecting to have attributes as : 
right 
left
operator 
or something like that cab u attach example?
tnx

**Reply by q-uint** [reply]
You would want to check the type of `Expression` in `params.Filter`.
For `userName eq Name` you would have an `AttributeExpression` with provides you with the attribute, operator and compare value.
